From 85ef725626b0ee446584f10de7b1ea65def6cbdd Mon Sep 17 00:00:00 2001
From: chris-belcher <chris-belcher@users.noreply.github.com>
Date: Mon, 16 Apr 2018 19:45:06 +0100
Subject: [PATCH] Email: fix crash logging in.

If user switches away from and then return to Email app when logging
into some pre-configured email account, a crash will occur. Logging
is done with AsyncTask. When it succeeds but the fragment is not in
resumed state, result is not delivered until next Fragment onResume()
call. It calls onCheckSettingsComplete() and then proceed(), involving
fragment transactions which is unsafe during FragmentManager state
changes. Delay UI updates to onPostResume().
---
 .../email/activity/setup/AccountCheckSettingsFragment.java     |  2 ++
 src/com/android/email/activity/setup/AccountSetupFinal.java    | 10 ++++++++++
 2 files changed, 12 insertions(+)

diff --git a/src/com/android/email/activity/setup/AccountCheckSettingsFragment.java b/src/com/android/email/activity/setup/AccountCheckSettingsFragment.java
index 1806b50..e22160c 100644
--- a/src/com/android/email/activity/setup/AccountCheckSettingsFragment.java
+++ b/src/com/android/email/activity/setup/AccountCheckSettingsFragment.java
@@ -174,7 +174,9 @@ public class AccountCheckSettingsFragment extends Fragment {
     public void onResume() {
         super.onResume();
         mPaused = false;
+    }
 
+    public void checkResult() {
         if (mState != STATE_START) {
             reportProgress(mState, mProgressException);
         }
diff --git a/src/com/android/email/activity/setup/AccountSetupFinal.java b/src/com/android/email/activity/setup/AccountSetupFinal.java
index adaa32a..3649a34 100644
--- a/src/com/android/email/activity/setup/AccountSetupFinal.java
+++ b/src/com/android/email/activity/setup/AccountSetupFinal.java
@@ -447,6 +447,16 @@ public class AccountSetupFinal extends AccountSetupActivity
     }
 
     @Override
+    protected void onPostResume() {
+        super.onPostResume();
+        AccountCheckSettingsFragment fragment = (AccountCheckSettingsFragment)
+                getFragmentManager().findFragmentByTag(AccountCheckSettingsFragment.TAG);
+        if (fragment != null) {
+            fragment.checkResult();
+        }
+    }
+
+    @Override
     public void onSaveInstanceState(@NonNull Bundle outState) {
         super.onSaveInstanceState(outState);
         outState.putBoolean(SAVESTATE_KEY_IS_PROCESSING, mIsProcessing);
-- 
2.11.0

